[jit] Add a utility class SourceRef to represent Source as keys

zhxchen17 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57398 [jit] Add Python API for ScriptProfile
* #57397 [jit] Implement ScriptProfile to collect instruction profiles.
* **#57396 [jit] Add a utility class SourceRef to represent Source as keys**

Summary:
A new type SourceRef is introduced to represent a unique identifier to source
text. The type holds refcount to underlying source, and supports comparators
and hash functions, such that it can be used in C++ and Python maps. In later
diffs we will use this to aggregate and print profiling information.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133578](https://our.internmc.facebook.com/intern/diff/D28133578)

zhxchen17 committed:
[jit] Add a utility class SourceRef to represent Source as keys

Summary:
A new type SourceRef is introduced to represent a unique identifier to source
text. The type holds refcount to underlying source, and supports comparators
and hash functions, such that it can be used in C++ and Python maps. In later
diffs we will use this to aggregate and print profiling information.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit] Add a utility class SourceRef to represent Source as keys"


Summary:
A new type SourceRef is introduced to represent a unique identifier to source
text. The type holds refcount to underlying source, and supports comparators
and hash functions, such that it can be used in C++ and Python maps. In later
diffs we will use this to aggregate and print profiling information.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

nikithamalgifb commented on torch/csrc/jit/frontend/source_ref.h:
```diff
@@ -0,0 +1,47 @@
+
+#include <c10/macros/Export.h>
+#include <torch/csrc/jit/frontend/source_range.h>
+
```
Can you add a small comment on what this file does ?

gmagogsfm commented on torch/csrc/jit/frontend/source_ref.h:
```diff
@@ -0,0 +1,47 @@
+  bool operator==(const SourceRef& other) const {
+    return source_ == other.source_;
+  }
+  bool operator<(const Source& other) const {
+    return source_.get() < &other;
+  }
+  friend bool operator<(const Source& other, const SourceRef& self) {
+    return &other < self.source_.get();
+  }
+  bool operator<(const SourceRef& other) const {
+    return *this < *other.source_.get();
+  }
```
Curious why these comparators are needed.

zhxchen17 replied:
I need to put this type as keys in std::map, and std::map is comparison based.

zhxchen17 committed:
addressed suggestions on "[jit] Add a utility class SourceRef to represent Source as keys"


Summary:
A new type SourceRef is introduced to represent a unique identifier to source
text. The type holds refcount to underlying source, and supports comparators
and hash functions, such that it can be used in C++ and Python maps. In later
diffs we will use this to aggregate and print profiling information.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133578](https://our.internmc.facebook.com/intern/diff/D28133578)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133578).

zhxchen17 committed:
nits on "[jit] Add a utility class SourceRef to represent Source as keys"


Summary:
A new type SourceRef is introduced to represent a unique identifier to source
text. The type holds refcount to underlying source, and supports comparators
and hash functions, such that it can be used in C++ and Python maps. In later
diffs we will use this to aggregate and print profiling information.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133578](https://our.internmc.facebook.com/intern/diff/D28133578)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133578).

zhxchen17 committed:
rebase on "[jit] Add a utility class SourceRef to represent Source as keys"


Summary:
A new type SourceRef is introduced to represent a unique identifier to source
text. The type holds refcount to underlying source, and supports comparators
and hash functions, such that it can be used in C++ and Python maps. In later
diffs we will use this to aggregate and print profiling information.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133578](https://our.internmc.facebook.com/intern/diff/D28133578)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133578).

zhxchen17 committed:
Update on "[jit] Add a utility class SourceRef to represent Source as keys"


Summary:
A new type SourceRef is introduced to represent a unique identifier to source
text. The type holds refcount to underlying source, and supports comparators
and hash functions, such that it can be used in C++ and Python maps. In later
diffs we will use this to aggregate and print profiling information.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133578](https://our.internmc.facebook.com/intern/diff/D28133578)

[ghstack-poisoned]

zhxchen17 committed:
Update on "[jit] Add a utility class SourceRef to represent Source as keys"


Summary:
A new type SourceRef is introduced to represent a unique identifier to source
text. The type holds refcount to underlying source, and supports comparators
and hash functions, such that it can be used in C++ and Python maps. In later
diffs we will use this to aggregate and print profiling information.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133578](https://our.internmc.facebook.com/intern/diff/D28133578)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133578).

zhxchen17 committed:
rebase on "[jit] Add a utility class SourceRef to represent Source as keys"


Summary:
A new type SourceRef is introduced to represent a unique identifier to source
text. The type holds refcount to underlying source, and supports comparators
and hash functions, such that it can be used in C++ and Python maps. In later
diffs we will use this to aggregate and print profiling information.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28133578](https://our.internmc.facebook.com/intern/diff/D28133578)

[ghstack-poisoned]

zhxchen17 commented:
@zhxchen17 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28133578).
